core/state_transition.go

```diff
@@ -471,31 +471,39 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
 }
 
 func (st *StateTransition) refundGas(refundQuotient uint64) uint64 {
-	// Apply refund counter, capped to a refund quotient
-	refund := st.gasUsed() / refundQuotient
-	if refund > st.state.GetRefund() {
-		refund = st.state.GetRefund()
+	// Calculate state change refund
+	stateRefund := st.gasUsed() / refundQuotient
+	if stateRefund > st.state.GetRefund() {
+		stateRefund = st.state.GetRefund()
 	}
 
-	if st.evm.Config.Tracer != nil && st.evm.Config.Tracer.OnGasChange != nil && refund > 0 {
-		st.evm.Config.Tracer.OnGasChange(st.gasRemaining, st.gasRemaining+refund, tracing.GasChangeTxRefunds)
+	// Calculate the total refund (unused gas + state refund)
+	totalRefund := st.gasRemaining + stateRefund
+
+	// Calculate the maximum refund allowed (150% of gas used)
+	maxRefund := st.gasUsed()/2 + st.gasUsed()
+
+	// Ensure the refund does not exceed the maximum allowed
+	if totalRefund > maxRefund {
+		totalRefund = maxRefund
 	}
 
-	st.gasRemaining += refund
+	// Update gasRemaining
+	st.gasRemaining += stateRefund
```

Comment:
it might be confusing to user if gasRemaining does not reflect the fact that their refund is capped. It might be clearer if we just set `st.gasRemaining = totalRefund` here

Reply:
sure

Reply:
I think the gasRemaining impacts the gas pool and gasUsed in the receipt. The refund cap should not affect the actual gasUsed and gasRemaining, but only impact the fee the user gets back.

Reply:
When set gasRemaining = totalRefund, the gasUsed in the transaction receipt is (gasWanted - totalRefund), not the actual gas used. For example, a transfer tx with --gas-limit=100,000, the receipt shows gasUsed as 68,500 instead of 21,000.

> it might be confusing to user if gasRemaining does not reflect the fact that their refund is capped. It might be clearer if we just set `st.gasRemaining = totalRefund` here

Reply:
that's true, makes sense